### ISSUE TITLE
Poset dimension SAT reduction

### ISSUE THREAD

**Issue by tiwwi** [sagemath/sage#35851]
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Problem Description

I recently came upon a SAT reduction for calculating the order dimension of a poset, that performs better than the one in sage, which I would like to contribute. This performs better than the current implementation in practice.

Instead of calculating a hypergraph colouring with potentially many edges, we calculate the order dimension by searching for transitive orientations of incomparable elements and introducing additional constraints such that each incomparable (ordered) pair is covered.

### Proposed Solution

I would like to implement this algorithm.

### Alternatives Considered

There is also an alternative formulation of the hypergraph colouring, which I have not tested yet.

### Additional Information

See also: [https://groups.google.com/g/sage-devel/c/uNhq1N3JhdI]( https://groups.google.com/g/sage-devel/c/uNhq1N3JhdI)

**Comment by dimpase** [comment]
Has the maths part been run by a human already, written up?

**Comment by tscrim** [comment]
@fchapoton

**Comment by tiwwi** [comment]
Sorry for the late response, I've been down with a cold the last couple of days (still am actually).

I've ran it past my advisor, but not in writing. I think we both considered it a pretty natural approach. I'll be having another meeting with him next week, if you prefer he checks it in more detail.

**Comment by tiwwi** [comment]
I have included the relevant pages of my thesis below, if you want to see the details. 
My thesis has now also been reviewed by my advisors and they agreed with this approach.

I have included the relevant pages from my thesis here: [sat_orders.pdf](https://github.com/sagemath/sage/files/14742181/sat_orders.pdf)

Just let me know if you want me to start up PR!

**Comment by dimpase** [comment]
Sure, PRs like this are very much welcome. We have our own issues with SAT implementations in Sage being either slow, or buggy, or both, however, so please be extra alert with these.
Do you use cryptominisat? (an optional package - should be faster than whatever built-in SAT solver we have). We have cryptominisat version 5.8.0 (from 2021), but the current one is 5.11.21.

I'll check whether this is easy to upgrade.

**Comment by tiwwi** [comment]
I was using the default SAT solver and kissat (which sage also interfaces with, if it is installed, I think). Personally, I didn't encounter any issues with SAT, so I'm reasonably optimistic that this should work out

**Comment by dimpase** [comment]
you wrote that Sage's dimension computation was slow - this could be in part due to the SAT solver

**Comment by dimpase** [comment]
@tiwwi - could you try an update of cryptominisat here: #37669 (and try your SAT computations with cryptominisat rather than other SAT solvers)?

**Comment by tiwwi** [comment]
@dimpase Yeah sure, I'll see if I can get this to work and then hopefully provide some more detailed benchmarks.
